hivemq-edge/src/main/java/com/hivemq/protocols/ConfigPersistence.java

```diff
@@ -0,0 +1,85 @@
+package com.hivemq.protocols;
+
+import com.hivemq.configuration.service.ConfigurationService;
+import com.hivemq.extension.sdk.api.annotations.NotNull;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * This class doesn't make use of caching in any way.
+ * Config is kept in memory so reading operations are fast.
+ *
+ * Beyond that this class is used to get all config interactions into one place for easier reqorks in the future.
+ */
+public class ConfigPersistence {
+
+    private static final Logger log = LoggerFactory.getLogger(ConfigPersistence.class);
+
+    private final @NotNull ConfigurationService configurationService;
+
+    public ConfigPersistence(final ConfigurationService configurationService) {
+        this.configurationService = configurationService;
+    }
+
+    public @NotNull Map<String, Object> allAdapters() {
+        return configurationService.protocolAdapterConfigurationService().getAllConfigs();
+    }
+
+    public void updateAllAdapters(final @NotNull Map<String, Object> adapterConfigs) {
+        configurationService.protocolAdapterConfigurationService().setAllConfigs(adapterConfigs);
+    }
+
+    public void addAdapter(final @NotNull String protocolId, final @NotNull Map<String, Object> config, final @NotNull List<Map<String, Object>> tagMaps) {
+        final Map<String, Object> mainMap = allAdapters();
+        final List<Map<String, Object>> adapterList = getAdapterListForType(mainMap, protocolId);
+        adapterList.add(combine(config, tagMaps));
+        updateAllAdapters(mainMap);
+    }
+
+    public void updateAdapter(final @NotNull String protocolId, final @NotNull Map<String, Object> config, final @NotNull List<Map<String, Object>> tagMaps) {
+        final Map<String, Object> mainMap = allAdapters();
+        final String adapterId = (String)config.get("id");
+        final List<Map<String, Object>> adapterList = getAdapterListForType(mainMap, protocolId);
```

Comment:
all these operations might throw ConcurrentModificationException if the map is not a concurrent one.

Reply:
Good catch.
Since I don't know I make a copy of the provided map into a HashMap when calling allAdapters.
I also synchronized all writing operations.